### ISSUE TITLE
☂️ [macOS] Support building engine on Apple Silicon without Rosetta

### ISSUE THREAD

**Issue by cbracken** [flutter/flutter#103386]
Currently, in order to build the engine on an Apple Silicon Mac, either as a host build or a target build, we need to build under Rosetta. This is for several reasons:

* [x] The Fuchsia clang toolchain for arm64 macOS is not yet available.
* [x] Swiftshader (used by our engine tests) doesn't support building as an arm64 binary. https://github.com/flutter/flutter/issues/90425

As such, we currently hardcode `host_cpu` to `x64` for all macOS engine builds (including cross-compiles targeting arm64), and as such these builds are run under Rosetta.

Once this work is done:
* [ ] We'll want to switch to the (host: arm64 target: arm64) `gen_snapshot` for arm64 macs rather than the current setup of running (host: x64 target: simarm64) `gen_snapshot` under Rosetta in the tool. We already support building on arm64 macs) rather than the (host: x64 target: simarm64) gen_snapshot.
* [x] There is no (host: arm64 target: simx64) `gen_snapshot` today since there's no support for this in the Dart SDK today. Until that work is prioritised, we'll need to continue generating x64 AOT code using the (host: x64 target: x64) `gen_snapshot` we use today, under Rosetta.
- [ ] Ask Chromium to remove Rosetta from Mac bots, reverting https://chrome-internal.googlesource.com/infra/puppet/+/6ed442c2c821fb2aeb0bf7f5f367d096b5da391a, see also https://bugs.chromium.org/p/chromium/issues/detail?id=1354817

Doing this is necessary to enable building unit tests in the host arm64 build on macOS. https://github.com/flutter/flutter/issues/115322 is the relevant sub-issue.

**Comment by manuelduarte077** [comment]
oh this is great.

**Comment by dnfield** [comment]
Also requires https://github.com/flutter/flutter/issues/90425

**Comment by cbracken** [comment]
I've got a patch WIP that switches us over to the clang arm64 toolchain on Apple Silicon macs rather than using Rosetta. We still need to use the x64 toolchain to build `gen_snapshot` variant targeting x64 macs.

**Comment by dkbast** [comment]
Are there any updates, or issues we could help with? I'm really looking forward to giving "this" (building the engine without rosetta) another try, so if there is anything I can help with, please let me know :)

**Comment by cosmicethics** [comment]
Hello, I'm a Flutter beginner. 
I want to build a macOS Flutter app specifically for the ARM architecture on arm-macos. I don't care about iOS/x64-macos and don't like rosetta. 
Are there any workarounds?

**Comment by TheSin-** [comment]
Did this fall off?  Seems there has been no work or update on this in some time.  Is there a arm64 or universal2 binary (using lipo) of the gen_snapshot_x binaries someplace we can manually install to help test with?

**Comment by karatekid430** [comment]
+1 seriously there are no more Intel Macs being made

**Comment by sumitsahoo** [comment]
Where are we on this?

**Comment by cbracken** [comment]
To post a quick update. 

Our top priority is getting Flutter _app developers_ using macOS hosts to the point where none of Flutter's tooling requires Rosetta. This work is tracked in https://github.com/flutter/flutter/issues/69157. Primarily this involves building universal binaries of `gen_snapshot` and other executables (`flutter_tester`, the `libimobiledevice` toolchain, etc.). Thus far, the macOS `gen_snapshot`s have been migrated (in https://github.com/flutter/engine/pull/53524 and https://github.com/flutter/engine/pull/53962) and iOS's `gen_snapshots` have been migrated (in https://github.com/flutter/engine/pull/54009). Android's `gen_snapshots` will be next, then `flutter_tester` and the `libimobiledevice` toolchain. This is the top priority since there are far more developers building apps with Flutter than there are developers who work on the engine.

Once all of that work is complete, we'll follow up with this issue, which eliminates the need for Rosetta for _engine developers_, unless _opting in_ to building with the x64 toolchain, which we still want to keep around to address issues with Intel hosts while working on Apple Silicon Macs.

**Comment by jmagman** [comment]
> * We'll want to switch to the (host: arm64 target: arm64) `gen_snapshot` for arm64 macs rather than the current setup of running (host: x64 target: simarm64) `gen_snapshot` under Rosetta in the tool. We already support building on arm64 macs) rather than the (host: x64 target: simarm64) gen_snapshot.

I'm trying to spot where this still happens after https://github.com/flutter/engine/pull/54291 and https://github.com/flutter/engine/pull/55882, maybe this is already done?
@zanderso since you did that work, do you remember where the (host: x64 target: simarm64) was/is still happening?

**Comment by jmagman** [comment]
> Our top priority is getting Flutter _app developers_ using macOS hosts to the point where none of Flutter's tooling requires Rosetta. This work is tracked in [#69157](https://github.com/flutter/flutter/issues/69157). 

The remaining work being tracked in this issue is NOT targeting Flutter developers, that is tracked in https://github.com/flutter/flutter/issues/69157. gen_snapshot, as installed on Flutter developer machines, has been a fat binary for awhile.

The remaining work in this issue is to build the ARM binaries on an ARM engine builder in CI without Rosetta, as opposed to the described (host: x64 target: simarm64) with Rosetta. From https://github.com/flutter/engine/pull/54291 and https://github.com/flutter/engine/pull/55882 I'm not seeing where we're building (host: x64 target: simarm64) any more, so maybe that work is completed.

I'm not sure how to actually test it without ripping Rosetta off all our engine builder bots and seeing what happens though.

**Comment by zanderso** [comment]
Before ripping Rosetta out of a test builder there are a couple of places we can audit for dependencies on Rosetta. We can  ensure that we aren't causing any intel binaries to show up on arm host by way of the CIPD artifacts referenced by:
- [ ] [The DEPS file](https://github.com/flutter/flutter/blob/master/DEPS)
- [ ] The [flutter_deps recipe module](https://flutter.googlesource.com/recipes/+/refs/heads/main/recipe_modules/flutter_deps/api.py)

As intermediate steps, we can adjust where/when the CIPD artifacts are downloaded on the builders.

**Comment by cbracken** [comment]
Alright, some good news. I've got a set of patches that produce successful engine builds for all target platforms supported by macOS hosts, without Rosetta on arm64 Macs. Everything has been tested in a VM without Rosetta installed.

All that said, there are a bunch of post-processing scripts we run as part of the Flutter release process (entitlements and code signing for example) that are not part of the regular build that developers do on their local machine, so I can't promise we're fully in the clear yet. 

Regardless, once these land, engine developers will no longer need Rosetta installed on their machines.

Patches I've sent out:
* https://github.com/flutter/flutter/pull/186660
* https://github.com/flutter/flutter/pull/186661
* https://github.com/flutter/flutter/pull/186662
* https://github.com/flutter/flutter/pull/186663
* https://github.com/flutter/flutter/pull/186664
* https://github.com/flutter/flutter/pull/186669

**Comment by cbracken** [comment]
As of 8740c456ec11b5e6185407297aa5935505146487, building any host or target build on an Apple Silicon macOS host machine should work just fine without Rosetta installed. I'd encourage anyone to test and provide bug reports if you run into problems.

**Comment by cbracken** [comment]
> FWIW: The real test is to just set up a machine from scratch and don't install rosetta and see if it works. :) Last time we built we had to install rosetta, but i think that was the 3.38.x series.

@eseidel were you thinking of something, like, say... this?

<img width="1136" height="912" alt="Image" src="https://github.com/user-attachments/assets/4e441256-f477-4655-adb5-3af937347de6" />

**Comment by cbracken** [comment]
As noted, at this point, then engine/embedder developer workflow should no longer require Rosetta.

Remaining work here is to verify work that happens on the bots that's not part of a typical engine/embedder developer's workflow -- e.g. entitlements and code signing and the rest of the release process. As @zanderso mentioned [above](https://github.com/flutter/flutter/issues/103386#issuecomment-3786303486), this includes auditing: 
- [x] [The DEPS file](https://github.com/flutter/flutter/blob/master/DEPS)
- [x] The [flutter_deps recipe module](https://flutter.googlesource.com/recipes/+/refs/heads/main/recipe_modules/flutter_deps/api.py)
- [x] Any cocoon or recipe-driven signing scripts/dart tooling. (TODO: replace this with detailed breakdown)

We'll also need a plan for:
* [x] Getting our hands on updated arm64 bot images without Rosetta.
* [x] How to test the builds on a test/beta pool devicelab bot without Rosetta.
* [x] How to test the release-build-specific functionality (code signing etc.) on the bots.
* [x] How to handle hotfixes on pre-de-rosettification branches once the arm64 bots are de-rosettified. I suspect my patches above apply cleanly to any release branches in the last ~year or so, but we'd want to validate that. Alternatively a plan for producing such builds ONLY on Intel bots.

/cc @jtmcdole for thoughts on the infra side.

**Comment by jtmcdole** [comment]
I love using Tart / Lume to validate this quickly. Thank you, @cbracken!

> How to test the builds on a test/beta pool devicelab bot without Rosetta.

We could run Lume on a bot so long as its M4 or higher. We have new M4's we're still trying to install in the devicelab.

**Comment by cbracken** [comment]
Alright our staging bots, trybots, and prod CI bots are all Rosetta-free. There's still one pool to eliminate Rosetta from, and then we'll need to do a beta and stable release to prove that the release workflow works, but we're really close here :)
